src/pykeen/evaluation/rank_based_evaluator.py

```diff
@@ -32,23 +52,107 @@
 
 logger = logging.getLogger(__name__)
 
-SIDE_HEAD = "head"
-SIDE_TAIL = "tail"
-SIDE_BOTH = "both"
-SIDES = {SIDE_HEAD, SIDE_TAIL, SIDE_BOTH}
-
-RANK_OPTIMISTIC = "optimistic"
-RANK_PESSIMISTIC = "pessimistic"
-RANK_REALISTIC = "realistic"
-RANK_TYPES = {RANK_OPTIMISTIC, RANK_PESSIMISTIC, RANK_REALISTIC}
-
-RANK_EXPECTED_REALISTIC = "expected_realistic"
-EXPECTED_RANKS = {
-    RANK_REALISTIC: RANK_EXPECTED_REALISTIC,
-    RANK_OPTIMISTIC: None,  # TODO - research problem
-    RANK_PESSIMISTIC: None,  # TODO - research problem
-}
+# typing
+Side = Literal["head", "tail"]
+ExtendedSide = Union[Side, Literal["both"]]
+# SIDE_HEAD, SIDE_TAIL = typing.get_args(Side) # Python >= 3.8
+SIDE_HEAD: Side = "head"
+SIDE_TAIL: Side = "tail"
+SIDE_BOTH: ExtendedSide = "both"
+
+# REAL_SIDES: Tuple[Side, ...] = typing.get_args(Side)  # Python >= 3.8
+REAL_SIDES: Tuple[Side, ...] = (SIDE_HEAD, SIDE_TAIL)
+SIDES: Tuple[ExtendedSide, ...] = cast(Tuple[ExtendedSide, ...], REAL_SIDES) + (SIDE_BOTH,)
+
+RankType = Literal["optimistic", "realistic", "pessimistic"]
+# RANK_TYPES: Tuple[RankType, ...] = typing.get_args(RankType) # Python >= 3.8
+RANK_TYPES: Tuple[RankType, ...] = ("optimistic", "realistic", "pessimistic")
+RANK_OPTIMISTIC, RANK_REALISTIC, RANK_PESSIMISTIC = RANK_TYPES
+
+
+@dataclass
+class ValueRange:
+    """A value range description."""
+
+    #: the lower bound
+    lower: Optional[float] = None
+
+    #: whether the lower bound is inclusive
+    lower_inclusive: bool = False
+
+    #: the upper bound
+    upper: Optional[float] = None
+
+    #: whether the upper bound is inclusive
+    upper_inclusive: bool = False
+
+    def __contains__(self, x: SupportsFloat) -> bool:
+        if self.lower is not None:
+            if x < self.lower:
+                return False
+            if not self.lower_inclusive and x == self.lower:
+                return False
+        if self.upper is not None:
+            if x > self.upper:
+                return False
+            if not self.upper_inclusive and x == self.upper:
+                return False
+        return True
+
+
+class RankBasedMetric:
+    """A base class for rank-based metrics."""
+
+    # TODO: verify interpretation
+    #: whether it is increasing, i.e., larger values are better
+    increasing: ClassVar[bool] = False
+
+    #: the value range (as string)
+    value_range: ClassVar[Optional[ValueRange]] = None
+
+    #: the supported rank types. Most of the time equal to all rank types
+    supported_rank_types: ClassVar[Collection[RankType]] = RANK_TYPES
+
+    #: synonyms for this metric
+    synonyms: ClassVar[Collection[str]] = tuple()
+
+    #: whether the metric requires the number of candidates for each ranking task
+    needs_candidates: ClassVar[bool] = False
+
+    @abstractmethod
+    def __call__(self, ranks: np.ndarray, num_candidates: Optional[np.ndarray] = None) -> float:
```

Comment:
will there ever be a need to have instance information (i.e., `self`)? Can the call be a static method? Only issue is you can't make `__call__` static on a class because python uses that to actually create the class itself

Reply:
e.g., Hits@k could need the `k` instance information.